File: web/src/lib/access.ts
// How far the nearest operating surgery is from a district that has no
// contracted physician (data/access.json). A distance is an accessibility
// proxy, never a claim that the district is unserved — substitution does not
// appear in the published data (CLAUDE.md rule 4).
import { useEffect, useState } from 'react';
import { t } from './i18n';
import type { ColDef, Row } from './eesztTable';
import type { PraxisKind } from '../types';

export interface AccessDistrict {
  id: string;
  kind: PraxisKind;
  status: string;
  settlement: string;
  county: string;
  type: string;
  population: number | null;
  km: number;
  band: string;
  nearestId: string;
  nearestSettlement: string;
  sameSettlement: boolean;
  geoApprox: boolean;
  lat: number;
  lon: number;
}

export interface AccessStats {
  subjects: number;
  measured: number;
  missingGeo: number;
  operating: number;
  operatingWithoutGeo: number;
  counts: Record<string, number>;
  population: Record<string, number>;
  medianKm: number | null;
  maxKm: number | null;
  sameSettlement: number;
}

export interface AccessRaw {
  schemaVersion: number;
  asOf: string;
  dataMonth: string;
  bands: { key: string; maxKm: number | null }[];
  stats: Record<PraxisKind, AccessStats>;
  districts: AccessDistrict[];
}

export const BAND_COLORS: Record<string, string> = {
  '0-5': '#4fd6c2',
  '5-10': '#c98500',
  '10-20': '#ff7a59',
  '20+': '#e05b8a',
};

let cache: AccessRaw | null = null;
let pending: Promise<AccessRaw | null> | null = null;

export function loadAccess(): Promise<AccessRaw | null> {
  if (cache) return Promise.resolve(cache);
  pending ??= fetch(`${import.meta.env.BASE_URL}data/access.json`)
    .then((r) => (r.ok ? r.json() : null))
    .then((d: AccessRaw | null) => { cache = d; return d; })
    .catch(() => null);
  return pending;
}

export function useAccess(): AccessRaw | null {
  const [data, setData] = useState<AccessRaw | null>(cache);
  useEffect(() => {
    let alive = true;
    void loadAccess().then((d) => { if (alive) setData(d); });
    return () => { alive = false; };
  }, []);
  return data;
}

export function bandLabel(key: string): string {
  return t(`access.band.${key}`);
}

export const ACCESS_COLUMNS: ColDef[] = [
  { key: 'fin', labelKey: 'eeszt.colFin', type: 'text', visible: true },
  { key: 'settlement', labelKey: 'stats.thSettlement', type: 'text', visible: true },
  { key: 'county', labelKey: 'stats.thCounty', type: 'enum', visible: true },
  { key: 'type', labelKey: 'stats.thType', type: 'enum', visible: false },
  { key: 'status', labelKey: 'stats.thStatus', type: 'enum', visible: true },
  { key: 'population', labelKey: 'access.colPopulation', type: 'number', visible: true },
  { key: 'km', labelKey: 'access.colKm', type: 'number', visible: true },
  { key: 'bandLabel', labelKey: 'access.colBand', type: 'enum', visible: true },
  { key: 'nearestSettlement', labelKey: 'access.colNearest', type: 'text', visible: true },
  { key: 'sameSettlement', labelKey: 'access.colSameSettlement', type: 'bool', visible: true },
  { key: 'nearestId', labelKey: 'access.colNearestId', type: 'text', visible: false },
  { key: 'geoApprox', labelKey: 'eeszt.colGeoApprox', type: 'bool', visible: false },
];

/** rows for the table and the map (the map colours by `type`) */
export function accessRows(data: AccessRaw | null, kind: PraxisKind): Row[] {
  if (!data) return [];
  return data.districts.filter((d) => d.kind === kind).map((d) => ({
    fin: d.id,
    settlement: d.settlement,
    county: d.county,
    type: bandLabel(d.band),          // the map's colour category
    praxisType: t(`praxisTypes.${d.type}`),
    status: d.status === 'dissolved' ? t('stats.statusDissolved') : t('stats.statusVacant'),
    population: d.population,
    km: d.km,
    bandLabel: bandLabel(d.band),
    nearestSettlement: d.nearestSettlement,
    sameSettlement: d.sameSettlement,
    nearestId: d.nearestId,
    geoApprox: d.geoApprox,
    lat: d.lat,
    lon: d.lon,
    unitCode: null,
    licenceId: null,
    providerId: null,
  }));
}

export function bandBreakdown(
  data: AccessRaw | null, kind: PraxisKind,
): { band: string; count: number; population: number }[] {
  const st = data?.stats?.[kind];
  if (!data || !st) return [];
  return data.bands.map((b) => ({
    band: b.key,
    count: st.counts[b.key] ?? 0,
    population: st.population[b.key] ?? 0,
  }));
}

/** people living in districts whose nearest operating surgery is beyond `km` */
export function populationBeyond(
  data: AccessRaw | null, kind: PraxisKind, km: number,
): number {
  if (!data) return 0;
  return data.districts
    .filter((d) => d.kind === kind && d.km >= km)
    .reduce((sum, d) => sum + (d.population ?? 0), 0);
}

/* ---------------- county summary ---------------- */

export interface CountyAccess {
  county: string;
  districts: number;
  medianKm: number;
  meanKm: number;
  maxKm: number;
  within5: number;
  beyond10: number;
  population: number;
  populationBeyond10: number;
}

function median(values: number[]): number {
  if (!values.length) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

/** per-county distances of the districts without a contracted physician */
export function countyAccess(data: AccessRaw | null, kind: PraxisKind): CountyAccess[] {
  if (!data) return [];
  const byCounty = new Map<string, AccessDistrict[]>();
  for (const d of data.districts) {
    if (d.kind !== kind) continue;
    const list = byCounty.get(d.county) ?? [];
    list.push(d);
    byCounty.set(d.county, list);
  }
  return [...byCounty.entries()]
    .map(([county, list]) => {
      const km = list.map((d) => d.km);
      return {
        county,
        districts: list.length,
        medianKm: Math.round(median(km) * 10) / 10,
        meanKm: Math.round((km.reduce((a, b) => a + b, 0) / km.length) * 10) / 10,
        maxKm: Math.round(Math.max(...km) * 10) / 10,
        within5: list.filter((d) => d.km < 5).length,
        beyond10: list.filter((d) => d.km >= 10).length,
        population: list.reduce((a, d) => a + (d.population ?? 0), 0),
        populationBeyond10: list
          .filter((d) => d.km >= 10)
          .reduce((a, d) => a + (d.population ?? 0), 0),
      };
    })
    .sort((a, b) => b.meanKm - a.meanKm);
}

export const COUNTY_ACCESS_COLUMNS: ColDef[] = [
  { key: 'county', labelKey: 'stats.thCounty', type: 'enum', visible: true },
  { key: 'districts', labelKey: 'access.colDistricts', type: 'number', visible: true },
  { key: 'meanKm', labelKey: 'access.colMeanKm', type: 'number', visible: true },
  { key: 'medianKm', labelKey: 'access.colMedianKm', type: 'number', visible: true },
  { key: 'maxKm', labelKey: 'access.colMaxKm', type: 'number', visible: true },
  { key: 'within5', labelKey: 'access.colWithin5', type: 'number', visible: true },
  { key: 'beyond10', labelKey: 'access.colBeyond10', type: 'number', visible: true },
  { key: 'population', labelKey: 'access.colPopulation', type: 'number', visible: false },
  { key: 'populationBeyond10', labelKey: 'access.colPopBeyond10', type: 'number', visible: true },
];

export function countyAccessRows(data: AccessRaw | null, kind: PraxisKind): Row[] {
  return countyAccess(data, kind).map((c) => ({ ...c }));
}

File: etl/access.py
"""How far the nearest operating surgery is from a district without a doctor.

A vacancy on its own says nothing about what a resident can actually reach:
an empty district whose neighbouring surgery is two kilometres away is a very
different situation from one where the nearest contracted physician is
twenty-five kilometres off. This module measures that distance for every
vacant and dissolved district:

    subject    = a district with no contracted physician (vacant / dissolved)
                 at its NEAK surgery address (geocoded)
    operating  = a FILLED district of the same branch, at its EESZT licensed
                 premises, or at its NEAK surgery address when the EESZT
                 match failed
    distance   = great-circle kilometres to the nearest operating surgery

This is an accessibility proxy, never a claim that the district is unserved
(CLAUDE.md rule 4): substitution is invisible in the published data, and the
crow-flies distance is not travel time. The output says so, and the UI repeats
it.

Usage:
  python etl/access.py
"""
from __future__ import annotations

import json
import math
import statistics
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "data" / "access.json"
SCHEMA_VERSION = 1

# the bands the UI colours and counts by (upper bound in km, None = above)
BANDS: list[tuple[str, float | None]] = [
    ("0-5", 5.0), ("5-10", 10.0), ("10-20", 20.0), ("20+", None),
]
EARTH_KM = 6371.0088


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp = p2 - p1
    dl = math.radians(lon2 - lon1)
    a = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
    return 2 * EARTH_KM * math.asin(math.sqrt(a))


def band_of(km: float) -> str:
    for key, upper in BANDS:
        if upper is None or km < upper:
            return key
    return BANDS[-1][0]


def build() -> dict:
    latest = json.loads((ROOT / "data" / "latest.json").read_text(encoding="utf-8"))
    eeszt_path = ROOT / "data" / "eeszt.json"
    eeszt = json.loads(eeszt_path.read_text(encoding="utf-8")) if eeszt_path.exists() else {"praxes": {}}
    cache = json.loads((ROOT / "etl" / "geocode_cache.json").read_text(encoding="utf-8"))
    eeszt_geo = eeszt.get("praxes", {})

    districts: list[dict] = []
    stats: dict[str, dict] = {}

    for kind in ("dental", "gp"):
        snap = latest["kinds"][kind]
        # where care is actually contracted: the filled districts
        operating: list[tuple[float, float, str, str]] = []
        no_geo = 0
        for f in snap["filledPraxes"]:
            entry = eeszt_geo.get(f["id"]) or {}
            geo = entry.get("g")
            if geo:
                operating.append((geo[0], geo[1], f["id"], f.get("settlement", "")))
                continue
            hit = cache.get(f"{f.get('postalCode', '')} {f.get('settlement', '')}, "
                            f"{f.get('address', '')}")
            if hit:
                operating.append((hit["lat"], hit["lon"], f["id"], f.get("settlement", "")))
            else:
                no_geo += 1

        counts: dict[str, int] = {key: 0 for key, _ in BANDS}
        population: dict[str, int] = {key: 0 for key, _ in BANDS}
        distances: list[float] = []
        missing = 0

        for p in snap["praxes"]:
            site = (p.get("sites") or [{}])[0]
            lat, lon = site.get("lat"), site.get("lon")
            if lat is None or lon is None or not operating:
                missing += 1
                continue
            best_km = math.inf
            best = ("", "")
            for olat, olon, oid, osett in operating:
                km = haversine(lat, lon, olat, olon)
                if km < best_km:
                    best_km, best = km, (oid, osett)
            km = round(best_km, 1)  # band from the published value, not the raw one
            band = band_of(km)
            counts[band] += 1
            population[band] += p.get("population") or 0
            distances.append(km)
            districts.append({
                "id": p["id"],
                "kind": kind,
                "status": p.get("status", "vacant"),
                "settlement": site.get("settlement", ""),
                "county": p.get("county", ""),
                "type": p.get("type", ""),
                "population": p.get("population"),
                "km": km,
                "band": band,
                "nearestId": best[0],
                "nearestSettlement": best[1],
                "sameSettlement": (site.get("settlement", "") or "?") == best[1],
                "geoApprox": bool(site.get("geoApprox")),
                "lat": lat,
                "lon": lon,
            })

        stats[kind] = {
            "subjects": len(snap["praxes"]),
            "measured": len(distances),
            "missingGeo": missing,
            "operating": len(operating),
            "operatingWithoutGeo": no_geo,
            "counts": counts,
            "population": population,
            "medianKm": round(statistics.median(distances), 1) if distances else None,
            "maxKm": round(max(distances), 1) if distances else None,
            "sameSettlement": sum(1 for d in districts
                                  if d["kind"] == kind and d["sameSettlement"]),
        }

    out = {
        "schemaVersion": SCHEMA_VERSION,
        "asOf": eeszt.get("asOf", ""),
        "dataMonth": latest.get("month", ""),
        "bands": [{"key": k, "maxKm": u} for k, u in BANDS],
        "stats": stats,
        "districts": districts,
    }
    guard(out, latest)
    return out


def guard(out: dict, latest: dict) -> None:
    """Every measured district must be accounted for, and stay name-free."""
    allowed = {"id", "kind", "status", "settlement", "county", "type", "population",
               "km", "band", "nearestId", "nearestSettlement", "sameSettlement",
               "geoApprox", "lat", "lon"}
    for d in out["districts"]:
        unknown = set(d) - allowed
        if unknown:
            raise ValueError(f"{d['id']}: unexpected fields {sorted(unknown)}")
        if d["status"] == "filled":
            raise ValueError(f"{d['id']}: a filled district is not a subject here")
        if not 0 <= d["km"] <= 300:
            raise ValueError(f"{d['id']}: implausible distance {d['km']} km")
        if d["band"] != band_of(d["km"]):
            raise ValueError(f"{d['id']}: band {d['band']} does not match {d['km']} km")
    for kind, st in out["stats"].items():
        total = len(latest["kinds"][kind]["praxes"])
        if st["measured"] + st["missingGeo"] != total:
            raise ValueError(
                f"{kind}: {st['measured']} measured + {st['missingGeo']} without "
                f"coordinates != {total} districts")
        if sum(st["counts"].values()) != st["measured"]:
            raise ValueError(f"{kind}: band counts do not add up to the measured set")


def main() -> None:
    out = build()
    OUT.write_text(json.dumps(out, ensure_ascii=False, separators=(",", ":")) + "\n",
                   encoding="utf-8")
    print(f"wrote {OUT}: {len(out['districts'])} districts without a doctor")
    for kind, st in out["stats"].items():
        bands = " ".join(f"{k}={v}" for k, v in st["counts"].items())
        print(f"  {kind}: median {st['medianKm']} km, max {st['maxKm']} km | {bands} "
              f"| operating {st['operating']} (no coords: {st['operatingWithoutGeo']})")


if __name__ == "__main__":
    main()
